### DISCUSSION TITLE
Mean distance

### DISCUSSION THREAD

**Discussion by MihalisSD** [gstavrinos/ai_msc_robotic_systems_assignments#38] [Q&A]
Το mean distance που μετραει την προσεγγιστικη απο την πραγματικη αποσταση των συντεταγμενων που εχει το μπαλακι ειναι σε μετρα;

**Comment by gstavrinos** [comment]
Ναι. Αν θεωρείς ότι μετράει λάθος πες μου να το κοιτάξω, δεν αποκλείεται!

**Comment by Xindictus** [comment]
Λίγο που το είδα κι εγώ, παίζει να μετράει λάθος δίχως να μαι απόλυτα σίγουρος. Πάντως πρόσεξα ότι το mean distance άλλαζε ενώ στο TF tree φαίνεται ακίνητο ή αν κινηθεί ελάχιστα (< 20cm) δείχνει μέχρι και 1m mean distance. Κοιτώντας λίγο καλύτερα τον υπολογισμό του mean distance, φαίνεται πως το υπολογίζει βάσει του frame ID `map` και όχι της κάμερας.

```c++
auto t = tf_buffer_->lookupTransform(latest_odom_->header.frame_id, latest_odom_->child_frame_id, tf2::TimePointZero);
```

**Reply by MihalisSD** [reply]
Παρατηρησα, αν χρησιμοποιεις το hough circles, οτι ανα τον χρονο αλλαζουν (κατα πολυ λιγο) οι μετρησεις που εξαγει ακομα και ακινητο να ειναι το μπαλακι. Μπορει να φταιει και αυτο.

**Reply by gstavrinos** [reply]
Φέρνει το transformation που παράγει ο κώδικάς σας στο frame που παράγεται από το simulation, και μετά τα συγκρίνει, οπότε, τελικά, δεν έχει σημασία ποιό είναι το frame.

Επίσης, όταν ψάχνεις κάτι το οποίο είναι <5cm, και το υπολογίζεις με ακρίβεια ~20cm, δεν το λες και ελάχιστη απόκλιση!

**Comment by MihalisSD** [comment]
Δεν μπορω να πω αν μετραει λαθος γιατι δεν ξερω αν οι τυπου που χρησιμοποιησα ειναι σωστοι αλλα νομιζω οτι η αποκλιση 0.5 αν ειναι σε μετρα ειναι σημαντικη.

![440327965_1117974096157592_3312214002674884138_n](https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/assets/151838345/8bf87719-21f8-4de4-8900-dec0e16edc15)

**Reply by gstavrinos** [reply]
Το 0.5 είναι σε μέτρα (όπως γράφει και στο output).

**Comment by gstavrinos** [comment]
Απαντάω γενικότερα για το mean distance εδώ.

Από αυτά που γράφετε, Μιχάλη και Κωνσταντίνε, φαίνεται πως έχετε σωστό κώδικα αλλά μειωμένη ακρίβεια. Το ότι η αυτόματη διόρθωση "κόβει" πολλούς βαθμούς για την ακρίβεια, μη σας απασχολεί τόσο. Αρκεί να είναι σωστή η λογική σας.

Αν ωστόσο πιστεύετε ότι υπάρχει κάποιο bug στο mean distance, το συζητάμε, εννοείται.

**Reply by Xindictus** [reply]
Πέρα από την ακρίβεια, που λογικά παίρνει βελτίωση αν και στο TF tree φαίνεται καλά τοποθετημένο το μπαλάκι, σε κάθε loop αυξάνεται πολύ το mean distance χωρίς να αλλάζει δραματικά έως και καθόλου το detection. Νομίζω έχει να κάνει με το παρακάτω:

```c++
mean_d_ += distance / dcnt_;
```

Αν καταλαβαίνω καλά, εμφανίζεις το mean distance ανά 10 μετρήσεις και χτίζεις ένα "incremental average". Σε αυτή την περίπτωση όμως δε θα έπρεπε να έχεις ένα temp `dsum` για τα distances και αυτό να διαιρείς με το `dcnt_`?

```c++
dcnt_++;
dsum_ += distance;
mean_d_ = dsum_ / dcnt_;
```

Καταλαβαίνω καλά;

**Reply by gstavrinos** [reply]
Έχεις δίκιο, πρόσθεσα το bug στο τελευταίο commit, προσπαθώντας να διορθώσω ένα παρόμοιο. Το διόρθωσα νομίζω, αλλά τα VMs δυστυχώς θα έχουν το πρόβλημα προς το παρόν. Στέλνω ανακοίνωση για pull. Thanks!

**Reply by Xindictus** [reply]
```shell
[tt_umpire-40] [INFO] [1715099069.631232903] [tt_umpire]: Mean distance = 0.055011 meters!
```

![yesyes-now-were](https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/assets/7975922/39d47a8f-52f8-42f1-844a-408f7241ef23)
